### PR TITLE
[GH_16717/MM_20389] Migrate 'components/signup/signup_controller' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by aggmoulik**
Signed-off-by: Moulik Aggarwal <qwertymoulik@gmail.com>

#### Summary
It includes the Migration to TypeScript and it should be merged after the redux changes.

#### Ticket Link
If this pull request addresses a Help Wanted ticket, 
  Fixes https://mattermost.atlassian.net/browse/MM-20389

#### Related Pull Requests

- Has server changes (please link here)
- Has redux changes [redux](https://github.com/mattermost/mattermost-redux/pull/1412)
- Has mobile changes (please link here)

### CODE REVIEW

**Comment by devinbinnie** [comment]
Hey @aggmoulik, thanks so much for the contribution!
I'm seeing seeing some TS compiler errors at this time, would you be able to fix them before I review? Let me know if you need help with any of them.

```
components/signup/signup_controller/index.ts:28:31 - error TS2339: Property 'NoAccounts' does not exist on type 'Partial<ClientConfig>'.

28     const noAccounts = config.NoAccounts === 'true';
                                 ~~~~~~~~~~

components/signup/signup_controller/signup_controller.test.tsx:59:14 - error TS2322: Type '{ location: { search: string; pathname: string; hash: string; }; noAccounts: boolean; loggedIn: boolean; isLicensed: boolean; isCloud: boolean; enableOpenServer: boolean; enableSAML: boolean; enableLDAP: boolean; ... 10 more ...; subscriptionStats: { ...; }; }' is not assignable to type 'Readonly<Props>'.
  Types of property 'subscriptionStats' are incompatible.
    Type '{ remainingSeats: number; isPaidTier: string; }' is missing the following properties from type 'SubscriptionStats': remaining_seats, is_paid_tier

59             <SignupController {...baseProps}/>,
                ~~~~~~~~~~~~~~~~

components/signup/signup_controller/signup_controller.test.tsx:71:14 - error TS2322: Type '{ isLicensed: boolean; location: { search: string; pathname: string; hash: string; }; noAccounts: boolean; loggedIn: boolean; isCloud: boolean; enableOpenServer: boolean; enableSAML: boolean; enableLDAP: boolean; ... 10 more ...; subscriptionStats: { ...; }; }' is not assignable to type 'Readonly<Props>'.
  Types of property 'subscriptionStats' are incompatible.
    Type '{ remainingSeats: number; isPaidTier: string; }' is not assignable to type 'SubscriptionStats'.

71             <SignupController {...props}/>,
                ~~~~~~~~~~~~~~~~

components/signup/signup_controller/signup_controller.test.tsx:95:14 - error TS2322: Type '{ location: { search: string; pathname: string; hash: string; }; actions: { addUserToTeamFromInvite: Mock<any, any>; getInviteInfo: Mock<any, any>; removeGlobalItem: Mock<any, any>; getTeamInviteInfo: Mock<...>; }; ... 16 more ...; subscriptionStats: { ...; }; }' is not assignable to type 'Readonly<Props>'.
  Types of property 'subscriptionStats' are incompatible.
    Type '{ remainingSeats: number; isPaidTier: string; }' is not assignable to type 'SubscriptionStats'.

95             <SignupController {...props}/>,
                ~~~~~~~~~~~~~~~~

components/signup/signup_controller/signup_controller.test.tsx:132:14 - error TS2322: Type '{ isCloud: boolean; location: { search: string; pathname: string; hash: string; }; subscriptionStats: { isPaidTier: string; remainingSeats: number; }; actions: { addUserToTeamFromInvite: Mock<any, any>; getInviteInfo: Mock<...>; removeGlobalItem: Mock<...>; getTeamInviteInfo: Mock<...>; }; ... 14 more ...; ldapLogin...' is not assignable to type 'Readonly<Props>'.
  Types of property 'subscriptionStats' are incompatible.
    Type '{ isPaidTier: string; remainingSeats: number; }' is missing the following properties from type 'SubscriptionStats': remaining_seats, is_paid_tier

132             <SignupController {...props}/>,
                 ~~~~~~~~~~~~~~~~

components/signup/signup_controller/signup_controller.test.tsx:158:14 - error TS2322: Type '{ location: { search: string; pathname: string; hash: string; }; actions: { addUserToTeamFromInvite: Mock<any, any>; removeGlobalItem: Mock<any, any>; getTeamInviteInfo: Mock<any, any>; }; ... 16 more ...; subscriptionStats: { ...; }; }' is not assignable to type 'Readonly<Props>'.
  Types of property 'subscriptionStats' are incompatible.
    Type '{ remainingSeats: number; isPaidTier: string; }' is not assignable to type 'SubscriptionStats'.

158             <SignupController {...props}/>,
                 ~~~~~~~~~~~~~~~~

components/signup/signup_controller/signup_controller.tsx:115:20 - error TS2339: Property 'isPaidTier' does not exist on type 'SubscriptionStats'.

115             const {isPaidTier, remainingSeats} = this.props.subscriptionStats;
                       ~~~~~~~~~~

components/signup/signup_controller/signup_controller.tsx:115:32 - error TS2339: Property 'remainingSeats' does not exist on type 'SubscriptionStats'.

115             const {isPaidTier, remainingSeats} = this.props.subscriptionStats;
```

**Comment by aggmoulik** [comment]
Hy @devinbinnie These errors can be solved after merging the redux PR. Will you be kind to check for QA then it will be a great help.

**Comment by devinbinnie** [comment]
> Hy @devinbinnie These errors can be solved after merging the redux PR. Will you be kind to check for QA then it will be a great help.

Ahh sorry, didn't notice that. My mistake. I've taken a look at that PR as well.

File: components/signup/signup_controller/signup_controller.tsx [hunk 1]
```diff
@@ -119,28 +136,32 @@ export default class SignupController extends React.PureComponent {
...
+                    loading: false,
+                });
+            } else if (error) {
+                this.handleInvalidInvite(error);
+            }
         }
     }
 
-    handleInvalidInvite = (err) => {
+    handleInvalidInvite = (err: { server_error_id: string; message: unknown }): void => {
```
Comment (devinbinnie) [inline components/signup/signup_controller/signup_controller.tsx:164]: Should this also be `ServerError`?

File: components/signup/signup_controller/signup_controller.tsx [hunk 2]
```diff
@@ -162,8 +183,8 @@ export default class SignupController extends React.PureComponent {
         });
     }
 
-    renderSignupControls = () => {
-        let signupControls = [];
+    renderSignupControls = (): unknown[] | void => {
```
Comment (devinbinnie) [inline components/signup/signup_controller/signup_controller.tsx:186]: Shouldn't this always return an array?

File: components/signup/signup_controller/signup_controller.tsx [hunk 3]
```diff
@@ -162,8 +183,8 @@ export default class SignupController extends React.PureComponent {
...
+        let signupControls: any = [];
```
Comment (devinbinnie) [inline components/signup/signup_controller/signup_controller.tsx:187]: This should probably be of type `React.ReactNode[]`

File: components/signup/signup_controller/signup_controller.tsx [hunk 4]
```diff
@@ -248,11 +269,13 @@ export default class SignupController extends React.PureComponent {
...
         if (this.props.isLicensed && this.props.enableSignUpWithOpenId) {
-            const buttonStyle = {};
+            const buttonStyle = {
+                backgroundColor: '',
+            };
             if (this.props.openidButtonColor) {
                 buttonStyle.backgroundColor = this.props.openidButtonColor;
             }
-            let buttonText = (
+            let buttonText: any = (
```
Comment (devinbinnie) [inline components/signup/signup_controller/signup_controller.tsx:278]: Use type `React.ReactNode` here

File: components/signup/signup_controller/signup_controller.tsx [hunk 5]
```diff
@@ -279,11 +302,11 @@ export default class SignupController extends React.PureComponent {
         }
 
         if (this.props.isLicensed && this.props.enableLDAP) {
-            const params = new URLSearchParams(this.props.location.search);
+            const params = new URLSearchParams(this.props?.location?.search);
             params.append('extra', 'create_ldap');
             const query = '?' + params.toString();
 
-            let LDAPText = (
+            let LDAPText: any = (
```
Comment (devinbinnie) [inline components/signup/signup_controller/signup_controller.tsx:309]: Same here

File: components/signup/signup_controller/signup_controller.tsx [hunk 6]
```diff
@@ -396,6 +419,11 @@ export default class SignupController extends React.PureComponent {
             signupControls = this.renderSignupControls();
         }
 
+        let locationSearch;
```
Comment (devinbinnie) [inline components/signup/signup_controller/signup_controller.tsx:422]: Can we avoid this block and just use the coalesce block below?

File: components/signup/signup_controller/signup_controller.tsx [hunk 7]
```diff
@@ -432,13 +460,14 @@ export default class SignupController extends React.PureComponent {
                             />
                             {' '}
                             <Link
-                                to={'/login' + this.props.location.search}
+                                to={'/login' + locationSearch}
```
Comment (devinbinnie) [inline components/signup/signup_controller/signup_controller.tsx:463]:
```suggestion
                                to={'/login' + (this.props.location?.search || '')}
```

**Review by devinbinnie — CHANGES_REQUESTED** [review]
Thanks for the PR @aggmoulik, I've added some comments for you to review.

**Review by lieut-data — APPROVED** [review]
Nothing to add beyond @devinbinnie's comments above -- let us know if we can help @aggmoulik :)

**Comment by aggmoulik** [comment]
Yeah, I just didn't get time from the office last week. So, will update on weekends.

File: components/signup/signup_controller/signup_controller.tsx [hunk 8]
```diff
@@ -21,44 +22,59 @@ import LoadingScreen from 'components/loading_screen';
...
-            removeGlobalItem: PropTypes.func.isRequired,
-            getTeamInviteInfo: PropTypes.func.isRequired,
-            addUserToTeamFromInvite: PropTypes.func.isRequired,
-        }).isRequired,
-    }
+export type Props = {
+    location?: {
+        search: string;
+    };
+    isLicensed?: boolean;
```
[the remark this answers is not in the document]
Comment (aggmoulik) [reply components/signup/signup_controller/signup_controller.tsx:29]: Actually I tried but I am getting error from the `index.ts` file as I have changed it to `React.ComponentClass` as they want everything to be optional. I was hoping if you may know about it.

File: components/signup/signup_controller/signup_controller.tsx [hunk 9]
```diff
@@ -72,7 +88,7 @@ export default class SignupController extends React.PureComponent {
             if (inviteId) {
                 loading = true;
             } else if (!this.props.loggedIn) {
-                usedBefore = props.usedBefore;
+                usedBefore = props?.usedBefore ? props.usedBefore : usedBefore;
```
[the remark this answers is not in the document]
Comment (aggmoulik) [reply components/signup/signup_controller/signup_controller.tsx:91]:
```suggestion
                usedBefore = props?.usedBefore ?? usedBefore;
```
I prefer to use `??` nullish coalescing operator as `| |` does check for falsy values like `0` , `false` etc.

File: components/signup/signup_controller/signup_controller.tsx [hunk 10]
```diff
@@ -21,44 +22,59 @@ import LoadingScreen from 'components/loading_screen';
...
+    ldapLoginFieldName?: string;
+    openidButtonText?: string;
+    openidButtonColor?: string;
+    subscriptionStats?: SubscriptionStats;
+    isCloud?: boolean;
+    loggedIn?: boolean;
+    actions?: {
+        removeGlobalItem: (name: string) => void;
+        getTeamInviteInfo: (inviteId: string) => {
+            data: any;
```
[the remark this answers is not in the document]
Comment (aggmoulik) [reply components/signup/signup_controller/signup_controller.tsx]: I couldn't find type from the `mattermost-redux`, can you please help me out here .

**Comment by M-ZubairAhmed** [comment]
file do not exists any more

**Pull request closed without merging**